skbio/stats/ordination/_utils.py

```diff
@@ -196,3 +196,62 @@ def f_matrix(E_matrix):
     col_means = E_matrix.mean(axis=0, keepdims=True)
     matrix_mean = E_matrix.mean()
     return E_matrix - row_means - col_means + matrix_mean
+
+
+def center_distance_matrix_optimized(distance_matrix):
+    """
+    # FYI: If the used distance was euclidean, pairwise distances
+    # needn't be computed from the data table Y because F_matrix =
+    # Y.dot(Y.T) (if Y has been centred).
+    # But since we're expecting distance_matrix to be non-euclidian,
+    # we do the following computation as per
+    # Numerical Ecology (Legendre & Legendre 1998)
+    """
+
+    return f_matrix_optimized(e_matrix_optimized(distance_matrix))
+
+
+def e_matrix_optimized(distance_matrix):
+    """
+    Compute E matrix from a distance matrix.
+    Squares and divides by -2 the input elementwise. Eq. 9.20 in
+    Legendre & Legendre 1998.
+    """
+    distance_matrix = np.array(distance_matrix).astype(np.float)
```

Comment:
Oh, this doesn't actually look like it's inplace... why not just pull `distance_matrix._data` and operate directly on that?

Reply:
isn't the type already ensured with a distance matrix?

Reply:
1. Agreed; I removed the redundant `np.array()` recast.
2. I also believed type to already have been ensured with a distance matrix. But the unit tests threw an error that was resolved by coercing types to `np.float` explicitly here.